File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  arr: T[],
  predicate: (item: T) => boolean,
  func: (item: T) => U
): U | undefined {
  const found = arr.find(predicate); // Find the first element matching the predicate.
  return found !== undefined ? func(found) : undefined; // Apply the function if found.
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let current = 1n; // Start with base^0.
  while (true) {
    yield current;
    current *= base; // Generate the next power.
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  const file = await open(filePath, "r");
  let count = 0;

  // Count lines that are non-empty and not comments.
  for await (const line of file.readLines()) {
    const trimmedLine = line.trim();
    if (trimmedLine.length > 0 && !trimmedLine.startsWith("#")) {
      count++;
    }
  }

  return count;
}

type Box = { kind: "Box"; width: number; length: number; depth: number };
type Sphere = { kind: "Sphere"; radius: number };
export type Shape = Box | Sphere;

export function volume(shape: Shape): number {
  if (shape.kind === "Sphere") {
    return (4 / 3) * Math.PI * Math.pow(shape.radius, 3); // Volume of a sphere.
  } else {
    return shape.width * shape.length * shape.depth; // Volume of a box.
  }
}

export function surfaceArea(shape: Shape): number {
  if (shape.kind === "Sphere") {
    return 4 * Math.PI * Math.pow(shape.radius, 2); // Surface area of a sphere.
  } else {
    // Surface area of a box.
    return 2 * (
      shape.width * shape.length +
      shape.width * shape.depth +
      shape.length * shape.depth
    );
  }
}

export interface BinarySearchTree<T> {
  size(): number; // Returns the number of nodes in the tree.
  insert(value: T): BinarySearchTree<T>; // Inserts a value and returns a new tree.
  contains(value: T): boolean; // Checks if the value exists in the tree.
  inorder(): Iterable<T>; // Produces elements in sorted order.
  toString(): string; // Returns a string representation of the tree.
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0; // An empty tree has no elements.
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty<T>(), new Empty<T>()); // Creates a new node.
  }

  contains(value: T): boolean {
    return false; // Empty trees contain no values.
  }

  *inorder(): Iterable<T> {
    // No elements to traverse.
  }

  toString(): string {
    return "()"; // Represents an empty tree.
  }
}

class Node<T> implements BinarySearchTree<T> {
  private readonly value: T; // The value stored in the node.
  private readonly left: BinarySearchTree<T>; // The left subtree.
  private readonly right: BinarySearchTree<T>; // The right subtree.

  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    this.value = value;
    this.left = left;
    this.right = right;
  }

  size(): number {
    return 1 + this.left.size() + this.right.size(); // Count the node and its subtrees.
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right); // Insert in the left subtree.
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value)); // Insert in the right subtree.
    }
    return this; // Ignore duplicates.
  }

  contains(value: T): boolean {
    if (value === this.value) {
      return true; // Found the value.
    } else if (value < this.value) {
      return this.left.contains(value); // Search in the left subtree.
    } else {
      return this.right.contains(value); // Search in the right subtree.
    }
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder(); // Traverse the left subtree.
    yield this.value; // Yield the current node's value.
    yield* this.right.inorder(); // Traverse the right subtree.
  }

  toString(): string {
    const leftTree = this.left instanceof Empty ? "" : this.left.toString();
    const rightTree = this.right instanceof Empty ? "" : this.right.toString();
    return `(${leftTree}${this.value}${rightTree})`; // Format as "(left value right)".
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining = function
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* Finds the first element in a list that satisfies a predicate and applies a function to it. *)
let first_then_apply arr predicate func =
  try Some (func (List.find predicate arr))
  with Not_found -> None

(* Generates an infinite sequence of powers of a given base. *)
let powers_generator base =
  let rec aux current =
    Seq.Cons (current, fun () -> aux (current * base))
  in
  aux 1

(* Counts non-empty, non-comment lines in a file. *)
let meaningful_line_count filename =
  let ic = open_in filename in
  let rec count_lines acc =
    try
      let line = input_line ic in
      let trimmed = String.trim line in
      if trimmed = "" || String.get trimmed 0 = '#' then
        count_lines acc
      else
        count_lines (acc + 1)
    with End_of_file ->
      close_in ic; acc
  in
  count_lines 0

(* Defines shapes and computes volume or surface area. *)
type shape =
  | Box of { width : float; length : float; depth : float }
  | Sphere of { radius : float }

let volume shape =
  match shape with
  | Box { width; length; depth } -> width *. length *. depth
  | Sphere { radius } -> (4.0 /. 3.0) *. Float.pi *. (radius ** 3.0)

let surface_area shape =
  match shape with
  | Box { width; length; depth } ->
      2.0 *. ((width *. length) +. (width *. depth) +. (length *. depth))
  | Sphere { radius } -> 4.0 *. Float.pi *. (radius ** 2.0)

(* Binary search tree implementation. *)
module BinarySearchTree = struct
  type 'a tree =
    | Empty
    | Node of { value : 'a; left : 'a tree; right : 'a tree }

  (* Inserts a value into the tree. *)
  let rec insert tree value =
    match tree with
    | Empty -> Node { value; left = Empty; right = Empty }
    | Node { value = v; left; right } ->
        if value < v then Node { value = v; left = insert left value; right }
        else if value > v then Node { value = v; left; right = insert right value }
        else tree

  (* Checks if a value exists in the tree. *)
  let rec contains tree value =
    match tree with
    | Empty -> false
    | Node { value = v; left; right } ->
        if value = v then true
        else if value < v then contains left value
        else contains right value

  (* Computes the number of nodes in the tree. *)
  let rec size tree =
    match tree with
    | Empty -> 0
    | Node { left; right; _ } -> 1 + size left + size right

  (* Produces an inorder sequence of tree values. *)
  let rec inorder tree =
    match tree with
    | Empty -> Seq.empty
    | Node { value; left; right } ->
        Seq.append (inorder left) (Seq.cons value (inorder right))

  (* Returns a string representation of the tree. *)
  let to_string tree =
    let rec aux = function
      | Empty -> "()"
      | Node { value; left; right } -> "(" ^ aux left ^ string_of_int value ^ aux right ^ ")"
    in
    aux tree
end
